## Exploring eBay Car Sales Data
For this project, I will be exploring a dataset of used cars from eBay Kleinanzeigen. The dataset was originally scraped and uploaded to Kaggle by user orgesleka. You can find the dataset  [here](https://data.world/data-society/used-cars-data).

### The data dictionary provided with the data is as follows:
dateCrawled - When this ad was first crawled. All field-values are taken from this date.

name - Name of the car.

seller - Whether the seller is private or a dealer.

offerType - The type of listing

price - The price on the ad to sell the car.

abtest - Whether the listing is included in an A/B test.

vehicleType - The vehicle Type.

yearOfRegistration - The year in which the car was first registered.

gearbox - The transmission type.

powerPS - The power of the car in PS.

model - The car model name.

kilometer - How many kilometers the car has driven.

monthOfRegistration - The month in which the car was first registered.

fuelType - What type of fuel the car uses.

brand - The brand of the car.

notRepairedDamage - If the car has a damage which is not yet repaired.

dateCreated - The date on which the eBay listing was created.

nrOfPictures - The number of pictures in the ad.

postalCode - The postal code for the location of the vehicle.

lastSeenOnline - When the crawler saw this ad last online

In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv('autos.csv', encoding='Windows-1252')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.isnull().sum()

dateCrawled               0
name                      0
seller                    0
offerType                 0
price                     0
abtest                    0
vehicleType            5095
yearOfRegistration        0
gearbox                2680
powerPS                   0
model                  2758
odometer                  0
monthOfRegistration       0
fuelType               4482
brand                     0
notRepairedDamage      9829
dateCreated               0
nrOfPictures              0
postalCode                0
lastSeen                  0
dtype: int64

## Initial Observations
1. Values in name column are unconventional and may require some cleaning
2. Values in price column are strings and should be converted to int/float for descriptive analysis 
3. Year of registration values are int values and should be changed to string values
4. Odometer readings should be converted to int values and column name should be changed
5. Month of registration values should be changed to month names
6. Some columns have null values of up to 20% of the entire dataset

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
maps = {
    'dateCrawled': 'date_crawled',
    'offerType': 'offer_type',
    'powerPS': 'power_ps',
    'vehicleType': 'vehicle_type',
    'yearOfRegistration': 'registration_year',
    'monthOfRegistration': 'registration_month',
    'fuelType': 'fuel_type',
    'notRepairedDamage': 'unrepaired_damage',
    'dateCreated': 'ad_created',
    'nrOfPictures': 'nr_of_pictures',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen'
}

In [7]:
autos.rename(maps, inplace=True, axis=1)

In [8]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

I changed the the naming conventions of the columns from camelcase to snakecase by using the Dataframe.map() method

## Inital Exploration and Cleaning
I changed the "price", "odometer" and "yearOfRegistration" columns to int, int and str respectively

I also changed the column names of "price" to "price_$" and "odometer" to "odometer_km"

In [10]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [11]:
# remove non-numeric characters and convert odomter and price columns to int
# convert registration_year column to string
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)

In [12]:
# rename odometer and price columns
autos.rename({'odometer':'odometer_km'}, inplace=True, axis=1)

In [13]:
#drop unnecessary columns
autos.drop(['nr_of_pictures', 'seller', 'offer_type'], axis=1, inplace=True)

In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

## Exploring the odometer and price columns 
Using maximum and minimum values to look for outliers (unrealistically low or high) values that we might want to remove.

In [15]:
# Number of unique values
autos['odometer_km'].unique().shape

(13,)

In [16]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

There are 13 unique values in this column. However, the middle and top quartile ranges are all at 150000km. This could mean that the values in this column represent ranges. For example, 5000km could represent any value between 0-5000, 10000km could represent any value between 5000-10000 and 150000km could represent any value >150000. 

I will use the value counts method to see how the values in these ranges are spread out

In [17]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

Based on the results given, we can see that over 60% of the values fall within the >150000km range. This most likely explains why the middle and top quartile ranges are 150000. Per this observation, it is safe to assume that there are no outliers in this column

In [18]:
print(autos['price'].unique().shape)
print(autos['price'].describe())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


There are definitely some outliers in this range because certain cars are priced of 

In [19]:
autos['price'].value_counts().sort_index(ascending=False).head(50)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
116000      1
115991      1
115000      1
114400      1
109999      1
105000      2
104900      1
99900       2
99000       2
98500       1
Name: price, dtype: int64

In [20]:
autos.loc[autos['price'] > 200000, ['name', 'price', 'odometer_km']].sort_values(by='price', ascending=False)

name     price  \
39705                        Tausch_gegen_gleichwertiges  99999999   
42221                                  Leasinguebernahme  27322222   
39377                        Tausche_volvo_v40_gegen_van  12345678   
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  12345678   
27371                                         Fiat_Punto  12345678   
24384                            Schlachte_Golf_3_gt_tdi  11111111   
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  11111111   
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  10000000   
47634                                        Ferrari_FXX   3890000   
7814                                         Ferrari_F40   1300000   
22947                         Bmw_530d_zum_ausschlachten   1234566   
43049                                      2_VW_Busse_T3    999999   
514                     Ford_Focus_Turnier_1.6_16V_Style    999999   
37585                                Volkswagen_Jetta_GT    999990   
36818                                        Porsche_991    350000   
14715                 Rolls_Royce_Phantom_Drophead_Coupe    345000   
34723  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...    299000   
35923  Porsche_911_Targa_Exclusive_Edition__1_von_15_...    295000   
12682  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...    265000   
47337                                    BMW_Z8_roadster    259000   
38299                                Glas_BMW_mit_Wasser    250000   
37840                                        Porsche_997    220000   

       odometer_km  
39705       150000  
42221        40000  
39377       150000  
47598       150000  
27371       150000  
24384       150000  
2897         50000  
11137       100000  
47634         5000  
7814         50000  
22947       150000  
43049       150000  
514         125000  
37585       150000  
36818         5000  
14715        20000  
34723       100000  
35923         5000  
12682         5000  
47337        20000  
38299         5000  
37840        30000

Most of the outliers are above the 350000 price range. While a few of them are actually worth their price, it is okay to drop these rows because most of the cars above this price are just not worth the stated price. Example: A Volkswagen Jetta costing 999990 and other cars prices at 123456789

In [21]:
# Remove rows above
autos = autos[autos['price'] < 400000]

In [22]:
autos['price'].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

### Observations
After further cleaning, 14 rows of data have been dropped.

In [23]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

## Exploring the date columns

In [24]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

First 10 characters represent a day

In [25]:
# Distribution of date_crawled
date_crawled = autos['date_crawled'].str[0:10]

# Day with the most crawls
date_crawled.value_counts(dropna=False, ascending=False)

2016-04-03    1934
2016-03-20    1891
2016-03-21    1874
2016-03-12    1838
2016-03-14    1831
2016-04-04    1824
2016-03-07    1798
2016-04-02    1770
2016-03-19    1745
2016-03-28    1742
2016-03-29    1707
2016-03-15    1699
2016-04-01    1690
2016-03-30    1681
2016-03-08    1663
2016-03-09    1660
2016-03-22    1645
2016-03-26    1624
2016-03-11    1624
2016-03-23    1619
2016-03-10    1606
2016-03-31    1595
2016-03-25    1587
2016-03-17    1575
2016-03-27    1552
2016-03-16    1475
2016-03-24    1455
2016-03-05    1269
2016-03-13     778
2016-03-06     697
2016-04-05     655
2016-03-18     653
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

In [26]:
# Find the date rang for 
date_crawled.value_counts(dropna=False).sort_index(ascending=True)

2016-03-05    1269
2016-03-06     697
2016-03-07    1798
2016-03-08    1663
2016-03-09    1660
2016-03-10    1606
2016-03-11    1624
2016-03-12    1838
2016-03-13     778
2016-03-14    1831
2016-03-15    1699
2016-03-16    1475
2016-03-17    1575
2016-03-18     653
2016-03-19    1745
2016-03-20    1891
2016-03-21    1874
2016-03-22    1645
2016-03-23    1619
2016-03-24    1455
2016-03-25    1587
2016-03-26    1624
2016-03-27    1552
2016-03-28    1742
2016-03-29    1707
2016-03-30    1681
2016-03-31    1595
2016-04-01    1690
2016-04-02    1770
2016-04-03    1934
2016-04-04    1824
2016-04-05     655
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

### Observation
Ads were crawled between 5th March 2016 and 7th April 2016
3rd April 2016 had the highest crawled count


In [27]:
# Distribution of ad_created
ad_created = autos['ad_created'].str[0:10]

In [28]:
# Day with the most ads
ad_created.value_counts(dropna=False, ascending=False)

2016-04-03    1946
2016-03-20    1893
2016-03-21    1884
2016-04-04    1842
2016-03-12    1830
              ... 
2015-09-09       1
2016-01-13       1
2015-06-11       1
2016-01-03       1
2015-12-05       1
Name: ad_created, Length: 76, dtype: int64

In [29]:
# date range for ad creation
ad_created.value_counts(dropna=False).sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
              ... 
2016-04-03    1946
2016-04-04    1842
2016-04-05     592
2016-04-06     163
2016-04-07      64
Name: ad_created, Length: 76, dtype: int64

The date range for ad creation is from 11th June 2015 to 7th April 2016. However, most of the adds were created in 2016 with 3rd April 2016 having the highest count of ads created.

In [30]:
last_seen = autos['last_seen'].str[0:10]

In [31]:
last_seen.value_counts(dropna=False).sort_index()

2016-03-05       54
2016-03-06      221
2016-03-07      268
2016-03-08      379
2016-03-09      492
2016-03-10      538
2016-03-11      626
2016-03-12     1190
2016-03-13      449
2016-03-14      640
2016-03-15      794
2016-03-16      822
2016-03-17     1396
2016-03-18      371
2016-03-19      787
2016-03-20     1035
2016-03-21     1036
2016-03-22     1079
2016-03-23      929
2016-03-24      978
2016-03-25      960
2016-03-26      848
2016-03-27      801
2016-03-28     1042
2016-03-29     1116
2016-03-30     1242
2016-03-31     1191
2016-04-01     1155
2016-04-02     1244
2016-04-03     1268
2016-04-04     1231
2016-04-05     6212
2016-04-06    11046
2016-04-07     6546
Name: last_seen, dtype: int64

Most of the ads from 2015 are no longer seen which means they have probably been sold or taken down. A lot of ads from the latter part of the timeframe are still up though. 

In [32]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

For the registration _year column, we can see that most of the used cars were registered in the mid-2000s. However there are some outliers like the min registration year being 1000 and max registration year being 9999

In [33]:
autos['registration_year'].value_counts(dropna=False).sort_index().head(50)

1000     1
1001     1
1111     1
1500     1
1800     2
1910     9
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     3
1951     2
1952     1
1953     1
1954     2
1955     2
1956     5
1957     2
1958     4
1959     7
1960    33
1961     6
1962     4
1963     9
1964    12
1965    17
1966    22
1967    27
1968    26
1969    19
1970    45
1971    27
1972    35
1973    25
1974    24
1975    19
1976    27
1977    22
1978    47
1979    35
1980    97
1981    30
1982    43
1983    53
Name: registration_year, dtype: int64

In [34]:
# remove cars cars with registration years outside the range of 1900-2016
autos = autos[autos['registration_year'].between(1900,2016)]

In [35]:
autos['registration_year'].describe()

count    48016.000000
mean      2002.806002
std          7.306212
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

I chose the range 1900-2016 because the earliest registration for a car in the dataset was in 1800 but the first car every made was in the year 1886. This makes the year 1800 invalid/inaccurate therefore, a range starting from 1900 will provide far more accurate results. Also, the most recent year when you look at the date_crawled and ad_created columns are 2016

After dealing with the outliers, we've succesfully removed 1,970 rows of invalid data

## Exploring Price by Brand

In [36]:
brands = autos['brand']

In [37]:
brands.unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [38]:
brands.value_counts()[0:10]

volkswagen       10185
bmw               5283
opel              5194
mercedes_benz     4579
audi              4149
ford              3350
renault           2274
peugeot           1418
fiat              1242
seat               873
Name: brand, dtype: int64

In [39]:
brands_count = brands.value_counts(normalize=True, dropna=False).sort_values(ascending=False)
print(brands_count[0:10])

volkswagen       0.212117
bmw              0.110026
opel             0.108172
mercedes_benz    0.095364
audi             0.086409
ford             0.069768
renault          0.047359
peugeot          0.029532
fiat             0.025866
seat             0.018181
Name: brand, dtype: float64


In [40]:
top_5_percent = brands_count[brands_count > .04]
print(top_5_percent)

volkswagen       0.212117
bmw              0.110026
opel             0.108172
mercedes_benz    0.095364
audi             0.086409
ford             0.069768
renault          0.047359
Name: brand, dtype: float64


In [49]:
from pprint import pprint

average_price = {}

# autos[autos['brand'] == 'bmw']['price']

# loop through the top 5% brands list
for brand in top_5_percent.index:
    brand_df = autos[autos['brand'] == brand]
    average_price[brand] = brand_df['price'].mean()

pprint(average_price)

{'audi': 9093.65003615329,
 'bmw': 8102.536248343744,
 'ford': 3652.095223880597,
 'mercedes_benz': 8485.239571958942,
 'opel': 2876.716403542549,
 'renault': 2395.4164467897976,
 'volkswagen': 5231.081983308787}


1. Audi, Mercedes Benz and BMW have the highest average price because they are luxury brands

2. Ford, Opel and Renault are make relatively cheap cars

3. Volkswagen cars are the most popular cars in out dataset. The cars they make are averagely neither cheap nor expensive

# Storing Aggregate Data in a DataFrame

## Using the groupby() method

I used this method to group data by brands and then calculate the average mileage for every car made by the brand on Ebay

In [51]:
groups = autos.groupby('brand')
average_mileage = {}

# loop through the top 5% brands list
for brand in top_5_percent.index:
    brand_df = groups.get_group(brand)
    average_mileage[brand] = brand_df['odometer_km'].mean()

pprint(average_mileage)

{'audi': 129287.78018799711,
 'bmw': 132431.38368351315,
 'ford': 124068.65671641791,
 'mercedes_benz': 130856.0821139987,
 'opel': 129223.14208702349,
 'renault': 128183.81706244503,
 'volkswagen': 128724.10407461954}


In [53]:
mean_mileage = pd.Series(average_mileage)
mean_price = pd.Series(average_price)
print(mean_mileage)
print(mean_price)

volkswagen       128724.104075
bmw              132431.383684
opel             129223.142087
mercedes_benz    130856.082114
audi             129287.780188
ford             124068.656716
renault          128183.817062
dtype: float64
volkswagen       5231.081983
bmw              8102.536248
opel             2876.716404
mercedes_benz    8485.239572
audi             9093.650036
ford             3652.095224
renault          2395.416447
dtype: float64


In [56]:
mean_df = pd.DataFrame(mean_price, columns=['mean_price'])
mean_df

mean_price
volkswagen     5231.081983
bmw            8102.536248
opel           2876.716404
mercedes_benz  8485.239572
audi           9093.650036
ford           3652.095224
renault        2395.416447

In [57]:
mean_df['mean_mileage'] = mean_mileage

In [59]:
mean_df

mean_price   mean_mileage
volkswagen     5231.081983  128724.104075
bmw            8102.536248  132431.383684
opel           2876.716404  129223.142087
mercedes_benz  8485.239572  130856.082114
audi           9093.650036  129287.780188
ford           3652.095224  124068.656716
renault        2395.416447  128183.817062

### Obersvations
The mean mileage for the top 5% of car brands is withing the 125000km-150000km range. However it is important to note that the data above is probably skewed due to the high number of cars with an odometer reading >150000km